include/wil/registry.h

```diff
@@ -17,16 +17,2855 @@
+        {
+            return ::wil::reg::get_value_nothrow(key, nullptr, value_name, &return_value);
+        }
+#endif
+
+#if defined(__WIL_OLEAUTO_H_)
+        /**
+         * \brief Reads a REG_SZ value under a specified key
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A wil::unique_bstr receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        inline HRESULT get_value_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, ::wil::unique_bstr& return_value) WI_NOEXCEPT
+        {
+            return_value.reset();
+            const reg_view_details::reg_view_nothrow regview{ key };
+            return regview.get_value(subkey, value_name, return_value);
+        }
+
+        /**
+         * \brief Reads a REG_SZ value under a specified key
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A wil::unique_bstr receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        inline HRESULT get_value_nothrow(HKEY key, _In_opt_ PCWSTR value_name, ::wil::unique_bstr& return_value) WI_NOEXCEPT
+        {
+            return ::wil::reg::get_value_nothrow(key, nullptr, value_name, return_value);
+        }
+
+        /**
+         * \brief Reads a REG_SZ value under a specified key
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A wil::unique_bstr receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        inline HRESULT get_value_string_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, ::wil::unique_bstr& return_value) WI_NOEXCEPT

```

Comment:
What's the motivation to have both `get_value_nothrow` and `get_value_string_nothrow` ? You know from the overload that the caller wants a string.

Reply:
it's part of how we layered the interface.

all 3 function types (get_value_nothrow, try_get_value_nothrow, get_value) can be templatized on the type to accept the data read from the registry. just super-generic (likely useful if someone is building template libraries consuming this template library) -- or nice and simple for users who don't feel the requirement to put in the value type.

then we supply versions that the user can specify the registry value type (one of the popular asks/requirements). get_value_dword_nothrow, get_value_qword_nothrow, get_value_string_nothrow, get_value_expanded_string_nothrow, get_value_multistring_nothrow.

the string functions are templatized on the string type (with std::wstring being default).

hope this makes sense :)